File: DataStructures/Tree/SwapNodes.py
#!/bin/python3

import os
import sys
from queue import Queue

sys.setrecursionlimit(2000)

#
# Complete the swapNodes function below.
#
class Node:
    def __init__(self, data, level): 
        self.data = data  
        self.left = None  
        self.right = None 
        self.level = level
        
    #
def swapNodes(root, level):
    if root is None:
        return
    elif (root.level % level) == 0:
        tmp = root.left
        root.left = root.right
        root.right = tmp
    
    swapNodes(root.left, level)
    swapNodes(root.right, level)
    
def inOrder(root):
    if root is not None:
        if root.left:
            inOrder(root.left)
        print(root.data, end = " ")
        inOrder(root.right)
        
if __name__ == '__main__':
    n = int(input())

    q = Queue()
    root = Node(1,1)
    q.put(root)
    level = 1 
    mainroot = Node(0,0)
    for t in range(n):
        root = q.get()
        l, r = input().split(' ')
        if root is not None:
            level = root.level+1
            if l != '-1':

                root.left = Node(l,level)
                q.put(root.left)
            else:
                root.left = None 
            if r != '-1':

                root.right = Node(r,level)
                q.put(root.right)
            else:
                root.right = None
        if t == 0:
            mainroot = root
    
    k  = int(input())


    for o in range(k):
        lev = int(input())
        swapNodes(mainroot, lev)
        inOrder(mainroot)
        print()
        
    

File: Data Structures Problems(C++)/Tree/SwapNodes.cpp
 /* This is not my solution!!!!
 This is a C++ solution of one of the hackerrankers. 
 It took me so long to complete the python solution so I just put a similar solution here in C++ since I was too lazy lol*/

#include <iostream>
#include <queue>
using namespace std;
typedef struct node {
    int data;
    int level;
    node* left;
    node* right;
}node;

node* createNode(int data, int level) {
    node* newNode = new node();
    newNode->data = data;
    newNode->level = level;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

void swapChildNodes(node* parent, int level) {
    if(parent == NULL)
        return;
    if(parent->level == level) {
        node* temp = parent->left;
        parent->left = parent->right;
        parent->right = temp;
        return;
    }
    swapChildNodes(parent->left, level);
    swapChildNodes(parent->right, level);
}

void inOrder(node* root) {
    if(root->left != NULL)
        inOrder(root->left);
    cout << root->data << " ";
    if(root->right != NULL)
        inOrder(root->right);
}

int main() {
    int N,T;
    cin >> N;
    queue<node*> nodeQueue;
    node* nullNode = NULL;
    node* root = createNode(1,1);
    nodeQueue.push(root);
    node* mainroot = root;
    int level;
    /* reading from STDIN and creating Tree */
    for(int i = 0; i < N; i++) {
        root = nodeQueue.front(); nodeQueue.pop();
        if(root != NULL) {
            int l, r;
            cin >> l >> r;
            level = root->level + 1;
            root->left = (l != -1) ? (createNode(l,level)) : (nullNode);
            root->right = (r != -1) ? (createNode(r,level)) : (nullNode);
            nodeQueue.push(root->left);
            nodeQueue.push(root->right);
        }
        else
            i--;
    }
    /* Swapping Child Nodes at respective levels */
    cin >> T;
    for(int i = 0; i < T; i++) {
        int K; cin >> K;
        for(int j = 1; j * K < level; j++) {
            swapChildNodes(mainroot, j * K);
        }
        /* Printing In order Traversal */
        inOrder(mainroot); cout << endl;
    }
    return 0;
}